glasbey.py

```diff
@@ -38,118 +45,227 @@ def finish(self):
             pass
 
 
-MAX = 256
-NUM_COLORS = MAX * MAX * MAX
-
-
-def generate_color_table():
-    """
-    Generate a lookup table with all possible RGB colors, encoded in
-    perceptually uniform CAM02-UCS color space.
-
-    Table rows correspond to individual RGB colors, columns correspond to J',
-    a', and b' components. The table is stored as a NumPy array.
-    """
-
-    widgets = ["Generating color table: ", Percentage(), " ", Bar(), " ", ETA()]
-    pbar = ProgressBar(widgets=widgets, maxval=(MAX * MAX)).start()
-
-    i = 0
-    colors = np.empty(shape=(NUM_COLORS, 3), dtype=float)
-    for r in range(MAX):
-        for g in range(MAX):
-            d = i * MAX
-            for b in range(MAX):
-                colors[d + b, :] = (r, g, b)
-            colors[d:d + MAX] = cspace_convert(
-                colors[d:d + MAX], "sRGB255", "CAM02-UCS"
-            )
-            pbar.update(i)
-            i += 1
-    pbar.finish()
-    return colors
-
-
-def generate_palette(
-    colors,
-    size,
-    base=None,
-    no_black=False,
-    lightness_range=None,
-    chroma_range=None,
-    hue_range=None,
-):
-    # Initialize palette with given base or white color
-    if base:
-        palette = [colors[i, :] for i in base]
-    else:
-        palette = [colors[-1, :]]  # white
-    # Exclude greys (values with low Chroma in JCh) and set lightness range,
-    if lightness_range is not None:
-        jch = cspace_convert(colors, "CAM02-UCS", "JCh")
-        colors = colors[
-            (jch[:, 0] >= lightness_range[0]) & (jch[:, 0] <= lightness_range[1]), :
-        ]
-    if chroma_range is not None:
-        jch = cspace_convert(colors, "CAM02-UCS", "JCh")
-        colors = colors[
-            (jch[:, 1] >= chroma_range[0]) & (jch[:, 1] <= chroma_range[1]), :
-        ]
-    if hue_range is not None:
-        jch = cspace_convert(colors, "CAM02-UCS", "JCh")
-        if hue_range[0] > hue_range[1]:
-            colors = colors[
-                (jch[:, 2] >= hue_range[0]) | (jch[:, 2] <= hue_range[1]), :
-            ]
+class Glasbey:
+    def __init__(self,
+                 base_palette=None,
+                 overwrite_base_palette: bool = False,
```

Comment:
What's the point of overwriting base palette?

Reply:
I found it useful in my use case, where my code may iteratively generates additional colors.

The parameter is optional and the feature takes only about 3 lines of code.

The advantage is that Glasbey saves the new colors in a format and place it understands perfectly. (Good responsibility-driven design, no?)

Reply:
I'm fine having this option, though I'm still not sure if I understand your motivation/use-case :)